Infrastructure for creating JITServer AOT cache records during compilation

AlexeyKhrabrov opened the pull request:
This PR adds infrastructure for requesting necessary information from the client JVM for creating JITServer AOT cache records, and caching pointers to these records in client session data. The APIs modified or added in this PR will be used by the AOT compiler to create and remember AOT cache records corresponding to relocation and validation records.

AlexeyKhrabrov commented:
@mpirvu This pull request is ready for review. Thanks.

AlexeyKhrabrov commented:
Updated to fix some issues in the `rememberClass()` implementation.

mpirvu reviewed (commented):
LGTM. I think that some of the new methods may benefit from comments explaining their intended purpose.
Regarding the name of the AOT cache, there is the danger of each client using a different name and the server having too many distinct cache. I am thinking that having a default name may partially circumvent that problem. Even so, we need some mechanism to prevent the proliferation of caches. This can be done in a separate PR.

AlexeyKhrabrov commented on runtime/compiler/control/JITServerHelpers.cpp:
```diff
@@ -651,49 +657,39 @@ JITServerHelpers::packRemoteROMClassInfo(J9Class *clazz, J9VMThread *vmThread, T
-   const auto &recv = stream->read<ClassInfoTuple>();
-   *classInfoTuple = std::get<0>(recv);
-   return romClassFromString(std::get<0>(*classInfoTuple), trMemory);
+   auto recv = stream->read<ClassInfoTuple>();
```
In this case there is no difference. `auto &recv` in the original code binds to the temporary tuple returned from `stream->read()` (extending the life time of the temporary to that of `recv`). In the new version, the tuple is simply returned into a named variable `recv`. There is no difference in terms of copying. I changed it to conform to the rest of the code base (e.g. `JITClientCompilationThread.cpp`).

AlexeyKhrabrov commented on runtime/compiler/compile/J9Compilation.hpp:
```diff
@@ -342,6 +342,9 @@ class OMR_EXTENSIBLE Compilation : public OMR::CompilationConnector
    TR::list<TR_OpaqueMethodBlock *>& getMethodsRequiringTrampolines() { return _methodsRequiringTrampolines; }
+
+   bool isAOTCacheStore() const { return _aotCacheStore; }
+   void setAOTCacheStore(bool store) { _aotCacheStore = store; }
```
It will be done (in a later PR) in `CompilationInfoPerThreadBase::wrappedCompile()` after creating the compilation object.

AlexeyKhrabrov commented on runtime/compiler/runtime/JITClientSession.hpp:
```diff
@@ -197,19 +194,21 @@ class ClientSessionData
       J9ConstantPool *_constantPool;
       uintptr_t _classFlags;
       uintptr_t _classChainOffsetIdentifyingLoader;
+      std::string _classNameIdentifyingLoader;
```
Yes, it adds a few dozen bytes per class. One thing we can do to reduce this memory overhead is to clear this string once `_aotCacheClassRecord` is initialized.

AlexeyKhrabrov commented on runtime/compiler/runtime/JITClientSession.cpp:
```diff
@@ -139,7 +138,7 @@ ClientSessionData::initializeUnloadedClassAddrRanges(const std::vector<TR_Addres
 
    if (!_unloadedClassAddresses)
-      _unloadedClassAddresses = new (PERSISTENT_NEW) TR_AddressSet(_persistentMemory, maxRanges);
+      _unloadedClassAddresses = new (_persistentMemory) TR_AddressSet(_persistentMemory, maxRanges);
```
`new (_persistentMemory->_persistentAllocator.get())` has to be used for classes that don't have `TR_ALLOC` or `TR_PERSISTENT_ALLOC` in their definition. My impression from looking at other code is that `new (persistentMemory)` is the one commonly used for classes that support this form, whereas `new (persistentAllocator)` is mostly used for classes that don't support the other form.

AlexeyKhrabrov commented:
> Regarding the name of the AOT cache, there is the danger of each client using a different name and the server having too many distinct cache. I am thinking that having a default name may partially circumvent that problem. Even so, we need some mechanism to prevent the proliferation of caches. This can be done in a separate PR.

There is already a default AOT cache name (an empty string) if it's not specified at the client. One thing we can do (in the future) is to delete AOT caches after a period of inactivity (i.e. if no clients have used a specific cache in a long time).

AlexeyKhrabrov commented:
I've addressed all the comments, and also made a small change: the well-known classes record is now also cached in the `SymbolValidationManager` object for easier access.

mpirvu commented:
jenkins test sanity plinuxjit,xlinuxjit,zlinuxjit jdk11

pshipton commented:
This doesn't compile on all platforms, see https://openj9-jenkins.osuosl.org/job/Pipeline-Build-Test-JDK8/170/

pshipton commented:
Reverted via https://github.com/eclipse-openj9/openj9/pull/13953 until we can fix the problems.

AlexeyKhrabrov commented:
Compilation errors should now be fixed in #13956.
